### PR TITLE
Question with images

### LINKED ISSUE #175: Complete the question generator with the last remaining types

### PR DESCRIPTION
**Pull request by FrancoisPog**
## Changes
closes #175

- [x] Create sql scripts to generate questions
- [x] Add new types in questions generator
- [x] Tests
- [x] Update docs 
- [x] Display images on client
- [x] Make the question with images playable

Light             |  Dark
:-------------------------:|:-------------------------:
Screen Shot 2021-03-16 at 23 29 14  |  Screen Shot 2021-03-16 at 23 42 17

Light             |  Dark
:-------------------------:|:-------------------------:
Screen Shot 2021-03-17 at 00 09 03  |  Screen Shot 2021-03-17 at 00 22 18

### CODE REVIEW

File: server/controllers/question.js [hunk 1]
```diff
@@ -72,6 +73,18 @@ const generatorInfosByType = {
...
+  11: {
+    filename: "question_IM.sql",
+    before: "",
+    createWording: () => `Quelle molécule a la structure chimique suivante ?`,
+    title: "1 structure chimique - 4 molécules",
+  },
+  12: {
+    filename: "question_MI.sql",
+    before: "",
+    createWording: (subject) => `Quelle structure chimique a la molécule '${subject}' ?`,
```
Comment (nathanaelhoun) [inline server/controllers/question.js:85]:
```suggestion
    createWording: (subject) => `Quelle structure chimique correspond à la molécule '${subject}' ?`,
```

File: server/controllers/question.js [hunk 2]
```diff
@@ -72,6 +73,18 @@ const generatorInfosByType = {
     createWording: (subject) => `Quelle interaction la molécule « ${subject} » a-t-elle ?`,
     title: "1 molécule - 4 interactions",
   },
+  11: {
+    filename: "question_IM.sql",
+    before: "",
+    createWording: () => `Quelle molécule a la structure chimique suivante ?`,
```
Comment (nathanaelhoun) [inline server/controllers/question.js:79]:
```suggestion
    createWording: () => `Quelle molécule possède la structure chimique suivante ?`,
```

File: client/src/components/quiz/Answers.js [hunk 3]
```diff
@@ -15,7 +15,7 @@ const AnswerBtn = ({ value, isRight, showResult, onClick }) => {
         onClick(value);
       }}
     >
-      {value}
+      {isImage ? <img src={value} alt="Answer" /> : value}
```
Comment (nathanaelhoun) [inline client/src/components/quiz/Answers.js:18]:
```suggestion
      {isImage ? <img src={value} alt="Image de la réponse" /> : value}
```

File: client/src/components/quiz/Question.js [hunk 4]
```diff
@@ -1,14 +1,17 @@
...
   return (
     <div id="quiz-question">
       <h2>
         Question {numero}
         {maxQuestion > 0 && <span>/{maxQuestion}</span>}
       </h2>
-      <h1>{text}</h1>
+      <h1>
+        {text}
+        {image && <img src={image} alt="Question subject" />}
```
Comment (nathanaelhoun) [inline client/src/components/quiz/Question.js:13]:
```suggestion
        {image && <img src={image} alt="Molécule du sujet de la question" />}
```

File: server/controllers/question.js [hunk 5]
```diff
@@ -72,6 +73,18 @@ const generatorInfosByType = {
...
+  11: {
+    filename: "question_IM.sql",
+    before: "",
+    createWording: () => `Quelle molécule possède la structure chimique suivante ?`,
+    title: "1 structure chimique - 4 molécules",
+  },
+  12: {
+    filename: "question_MI.sql",
+    before: "",
+    createWording: (subject) => `Quelle structure chimique correspond à la molécule '${subject}' ?`,
```
Comment (nathanaelhoun) [inline server/controllers/question.js:85]:
```suggestion
    createWording: (subject) => `Quelle structure chimique correspond à la molécule « ${subject} » ?`,
```

File: server/controllers/question.js [hunk 6]
```diff
@@ -9,6 +9,7 @@ import { queryPromise } from "../db/database.js";
 // eslint-disable-next-line no-unused-vars
 import { HttpResponseWrapper } from "../global/HttpControllerWrapper.js";
 
+const IMAGES_ROUTE = "files/images";
 /* eslint-disable no-irregular-whitespace */
```
Comment (nathanaelhoun) [inline server/controllers/question.js:13]: I should have added it in my PR, sorry
```suggestion
/* eslint-disable no-irregular-whitespace */ // For non-breaking spaces around French quotes
```

File: server/controllers/question.js [hunk 7]
```diff
@@ -72,6 +73,18 @@ const generatorInfosByType = {
...
+    filename: "question_IM.sql",
+    before: "",
+    createWording: () => `Quelle molécule possède la structure chimique suivante ?`,
+    title: "1 structure chimique - 4 molécules",
+  },
+  12: {
+    filename: "question_MI.sql",
+    before: "",
+    createWording: (subject) => `Quelle structure chimique correspond à la molécule '${subject}' ?`,
+    title: "1 molécules - 4 structure chimique",
```
Comment (nathanaelhoun) [inline server/controllers/question.js:86]:
```suggestion
    title: "1 molécule - 4 structures chimiques",
```

File: server/test/question.test.js [hunk 8]
```diff
@@ -99,6 +117,19 @@ describe("Question generation", function () {
     );
   });
 
+  it("Type 4 : Consistent values", async () => {
```
Comment (nathanaelhoun) [inline server/test/question.test.js:120]:
```suggestion
  it("Type 4: Consistent values", async () => {
```

File: server/test/question.test.js [hunk 9]
```diff
@@ -99,6 +117,19 @@ describe("Question generation", function () {
...
+    const answersContainingSubject = await Promise.all(
+      answers.map((value) => doesBelongToSystem(subject, value))
+    );
+
+    answersContainingSubject.forEach((value, index) =>
+      expect(value).to.be.equals(index === Number(goodAnswer))
+    );
+  });
+
   it("Type 5 : Consistent values", async () => {
```
Comment (nathanaelhoun) [inline server/test/question.test.js:133]:
```suggestion
  it("Type 5: Consistent values", async () => {
```

File: server/test/question.test.js [hunk 10]
```diff
@@ -164,6 +195,39 @@ describe("Question generation", function () {
     );
   });
 
+  it("Type 10 : Consistent values", async () => {
```
Comment (nathanaelhoun) [inline server/test/question.test.js:198]:
```suggestion
  it("Type 10: Consistent values", async () => {
```

File: server/test/question.test.js [hunk 11]
```diff
@@ -164,6 +195,39 @@ describe("Question generation", function () {
...
+    const answersHavePropertyValue = await Promise.all(
+      answers.map((value) => doesHavePropertyValue(subject, "interactions", value))
+    );
+
+    answersHavePropertyValue.forEach((value, index) =>
+      expect(value).to.be.equals(index === Number(goodAnswer))
+    );
+  });
+
+  it("Type 11 : Good image url", async () => {
```
Comment (nathanaelhoun) [inline server/test/question.test.js:211]:
```suggestion
  it("Type 11: Good image url", async () => {
```

File: server/test/question.test.js [hunk 12]
```diff
@@ -164,6 +195,39 @@ describe("Question generation", function () {
...
+    const {
+      body: { subject },
+    } = await requestAPI("question/11");
+
+    const res = await requestAPI(subject.split("/api/v1/")[1]);
+    expect(res.status).equal(200);
+  });
+
+  it("Type 12 : Good image url", async () => {
```
Comment (nathanaelhoun) [inline server/test/question.test.js:220]:
```suggestion
  it("Type 12: Good image url", async () => {
```

**Review by nathanaelhoun — APPROVED** [review]
Great integration of the images!

Some non-blocking comments on the wordings :wink:

File: client/src/components/quiz/Answers.js [hunk 13]
```diff
@@ -15,7 +15,7 @@ const AnswerBtn = ({ value, isRight, showResult, onClick, isImage }) => {
         onClick(value);
       }}
     >
-      {isImage ? <img src={value} alt="Answer" /> : value}
+      {isImage ? <img src={value} alt="Image de la réponse" /> : value}
```
Comment (FrancoisPog) [inline client/src/components/quiz/Answers.js:18]:
```suggestion
      {isImage ? <img src={value} alt="Réponse" /> : value}
```

**Pull request merged**